### ISSUE TITLE
Add a tidy check that every method implemented in FooMethods traits has a comment containing a spec link

### ISSUE THREAD

**Issue by jdm** [servo/servo#6664]
We encourage this but forget to require it in reviews. It would be nice to have automation checking this for us. We might be able to use a heuristic like:
* for each file in components/script/dom/*.rs, find any line matching `impl<'a> [filename]Methods for &'a [filename] {`
* if it exists, set the nested brace count to 1, then read lines until the number of nested braces returns to 0 (ie. terminating } is found)
* from those lines, find each line like `fn [name](` where the nested brace count is 1
* look at the immediately preceding lines that contain `//` and error if there is no line containing `https://`

Code: python/tidy.py

**Comment by boghison** [comment]
I'd like to take this!

**Comment by boghison** [comment]
It will be easier for me to work if #6648 is merged, because I would like to avoid conflicts :)

**Comment by jdm** [comment]
Good point. I'll deal with that PR tomorrow!

**Comment by boghison** [comment]
I am having some trouble with adding links, for example with htmlbodyelement.rs. bgcolor is obsolete, therefore the current link doesn't work, and I can't find one. What should I do with GetOnunload and SetOnunload?

**Comment by jdm** [comment]
The best we can do is link to https://html.spec.whatwg.org/multipage/semantics.html#the-body-element, unfortunately.

**Comment by jdm** [comment]
The bgcolor link can be to https://html.spec.whatwg.org/multipage/obsolete.html#non-conforming-features

**Comment by jdm** [comment]
Actually, https://html.spec.whatwg.org/multipage/#dom-body-bgcolor is a better link for that.

**Comment by boghison** [comment]
What about onload?

**Comment by jdm** [comment]
For body? Same link to the body element as above.

**Comment by boghison** [comment]
What about bgcolor for document?

**Comment by jdm** [comment]
https://html.spec.whatwg.org/multipage/#dom-document-bgcolor

**Comment by boghison** [comment]
What should I do about workernavigator?

**Comment by boghison** [comment]
And testbinding(proxy)?

**Comment by jdm** [comment]
As for workernavigator, you should be able to find everything you need at https://html.spec.whatwg.org/multipage/#the-workernavigator-object .

Oh, the testbinding one is a test that has no corresponding spec. Maybe add the ability to opt out via a comment at the top of the file?

**Comment by boghison** [comment]
Like this:
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
//NO SPEC TEST

**Comment by jdm** [comment]
I'd say `// check-tidy: no spec`

**Comment by boghison** [comment]
@jdm I'll make it local to this function, but this could be interesting as an enhancement. Could you file an issue?

**Comment by boghison** [comment]
So I have this left and I'd like some more help:
components/script/dom/htmlcollection.rs:232: no spec link was found for the respective function
components/script/dom/htmlcollection.rs:238: no spec link was found for the respective function
components/script/dom/performancetiming.rs:42: (much) overlong line
components/script/dom/domtokenlist.rs:85: no spec link was found for the respective function
components/script/dom/storage.rs:85: no spec link was found for the respective function
components/script/dom/storage.rs:103: no spec link was found for the respective function
components/script/dom/storage.rs:107: no spec link was found for the respective function
components/script/dom/storage.rs:122: no spec link was found for the respective function
components/script/dom/window.rs:366: no spec link was found for the respective function
components/script/dom/window.rs:370: no spec link was found for the respective function
components/script/dom/window.rs:463: no spec link was found for the respective function
components/script/dom/window.rs:474: no spec link was found for the respective function
components/script/dom/window.rs:478: no spec link was found for the respective function
components/script/dom/window.rs:483: no spec link was found for the respective function
components/script/dom/window.rs:489: no spec link was found for the respective function
components/script/dom/window.rs:499: no spec link was found for the respective function
components/script/dom/window.rs:521: no spec link was found for the respective function
components/script/dom/window.rs:529: no spec link was found for the respective function
components/script/dom/workerglobalscope.rs:167: no spec link was found for the respective function
components/script/dom/domrectlist.rs:51: no spec link was found for the respective function
components/script/dom/nodelist.rs:76: no spec link was found for the respective function
components/script/dom/console.rs:43: no spec link was found for the respective function
components/script/dom/formdata.rs:110: no spec link was found for the respective function
components/script/dom/namednodemap.rs:94: no spec link was found for the respective function
components/script/dom/namednodemap.rs:100: no spec link was found for the respective function

In short, what should I do with the long line and what should I do with IndexedGetter, NamedGetter etc?

**Comment by jdm** [comment]
The specs talk about "supported property indices" for things like IndexedGetter/IndexedSetter/NamedGetter/NamedSetter. For example, https://html.spec.whatwg.org/multipage/#accessing-other-browsing-contexts .

I'm not sure why you have a long line error that wasn't there before. Are you using this style?
```rust
// http://spec.link.here
fn Whatever() {
}
```

**Comment by jdm** [comment]
Some like HTMLCollection are in the DOM spec, like https://dom.spec.whatwg.org/#dom-htmlcollection-nameditem .

**Comment by boghison** [comment]
It is long because:
    // https://dvcs.w3.org/hg/webperf/raw-file/tip/specs/NavigationTiming/Overview.html#dom-performancetiming-navigationstart

**Comment by boghison** [comment]
Should I use an url shortener?

**Comment by boghison** [comment]
That is not the same function.

**Comment by jdm** [comment]
Hah! I would just split it on two lines for now.

You can also feel free to recognise a special comment to allows methods to be missing a spec link, and we can fill them in later for the ones that are more difficult.

**Comment by boghison** [comment]
That won't work, because I am checking the previous line. I could just make the missing functions last, and then use the comment we talked about before.

**Comment by boghison** [comment]
I still need:
 - Console -> Debug
 - WorkerGlobalScope -> Console
 - Window -> Console; Crypto; Performace; Screen; Debug; Gc; Trap; WebdriverCallback; WebdriverTimeout

**Comment by jdm** [comment]
https://dvcs.w3.org/hg/webperf/raw-file/tip/specs/NavigationTiming/Overview.html#sec-window.performance-attribute
debug, gc, trap, webdrivercallback, and webdriver timeout have no specs. They're all nonstandard additions by Servo.
For console, link to https://developer.mozilla.org/en-US/docs/Web/API/Console and https://developer.mozilla.org/en-US/docs/Web/API/WorkerGlobalScope/console

**Comment by boghison** [comment]
@jdm Screen? Crypto?

**Comment by boghison** [comment]
Ok, I solved everything except the long line. If you want to, you can take a look at my commits in the tidyspec branch.

**Comment by jdm** [comment]
Exciting! I will note that many of the spec links are pointing at incorrect or outdated versions of the HTML5 spec, though. https://html.spec.whatwg.org/multipage/ is the canonical HTML5 spec, and you can see other appropriate ones in https://github.com/servo/servo/wiki/Relevant-spec-links .

**Comment by boghison** [comment]
@jdm Why was this closed?

**Comment by jdm** [comment]
Oops, I thought the PR had merged already. My mistake.

**Comment by jdm** [comment]
For future reference the PR is #6699.
